### ISSUE TITLE
Installation issue:  hdf +external-xdr ^libtirpc failing (libtirpc issue???) 

### ISSUE THREAD

**Issue by payerle** [LLNL/spack#23328]
I am attempting to build hdf on a 64 bit system using libtirpc for external-xdr, but the build fails complaining that the hdf configure parameter --enable-hdf4-xdr is not supported for 64 bit mode.

(This is a re-issue of #23297 --- I am in the process of moving from spack 0.14 to the most recent 0.16.1, and somehow the spack environment used in 23297 was corrupted.  I have cleaned up, confirmed that what I am using is a recent clone of repo, and that the problem persists).
Debugging the hdf configure script a bit further, it appears that spack is correctly not passing either --enable-hdf4-xdr or --disable-hdf4-xdr (due to a bug in hdf configure script the latter is treated the same as the former), and that the hdf configure script attempts to determine if an external xdr is available by compiling a test code, which fails.

The test code appears to be failing because it cannot find the libtirpc header files types.h, xdr.h, which libtirpc installs into
``` `spack location -i libtirpc`/include/tirpc/rpc```
but the include directory spack is using is only
``` `spack location -i libtirpc`/include ```, i.e. stops at the include directory. (I.e., the value of self.spec['rpc'].headers.cpp_flags inside the hdf recipe is "-I\$LIBTIRPC_PREFIX/include" and not -I\$LIBTIRPC_PREFIX/include/tirpc). This is causing issues because the hdf test code is doing a simple
```#include <rpc/xdr.h> ``` as can be seen in the attached config.log (further down).

I  believe this to be a bug in the spack recipe for libtirpc, as not only does the test code usage seem correct, but even the types.h file from libtirpc has include statements like
```#include <netconfig.h>```
which would also require the -I flags to include the tirpc subdirectory in order to work.

As a test, I manually added symlinks in the libtirpc installation directory
```console
cd `spack location -i libtircp`
cd include
ln tircp/* .
```
and the spack installation of hdf succeeded.

My old #15993 patch explicitly added ```-I\$LIBTIRPC_PREFIX/include/tirpc``` to the preprocessor flags, since then things have generalized into a virtual rpc dependency, but the extra 'tirpc' subdirectory got lost at some point, and without that the various rpc header files are not being found with an ```#include <rpc/xdr.h>``` or similar include directive.

So I believe the proper solution is to modify libtirpc to set compilation flags which include ```-I\$LIBTIRCP_PREFIX/include/tirpc```, possibly in addition to 
```-I\$LIBRIRPC_PREFIX/include```, but I am not sure how to achieve that in spack.  Alternatively, we could add special code to hdr to add an ```-I\$LIBTIRPC_PREFIX/include/tirpc`` flag  if it is using external-xdr and using libtirpc as the rpc provider, but this seems like it is likely to be needed in everything that possibly uses libtirpc for rpc, and so seems poor spack design.
  
### Steps to reproduce the issue

```console
$ spack install hdf@4.2.14 +external-xdr +fortran ~java ~netcdf +pic ~shared +szip %gcc@8.4.0
...
configure: error: --enable-hdf4-xdr is not supported for 64 bit mode.
```
where we are using libtirpc as the provider for 'rpc'

### Information on your system

```console$ spack debug report
* **Spack:** 0.16.1-2430-47196f30d7
* **Python:** 3.6.8
* **Platform:** linux-rhel8-x86_64
* **Concretizer:** original
```
As mentioned previously, our configuration includes
```console
packages:
  #Default settings for all packages
  all:
    buildable: true
    providers::
      ...
      rpc: [libtirpc]
     ...
```

### Additional information

* [spack-configure-args.txt](https://github.com/spack/spack/files/6393800/spack-configure-args.txt)
* [spack-build-out.txt](https://github.com/spack/spack/files/6393801/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/6393802/spack-build-env.txt)
* [spack-build-02-configure-out.txt](https://github.com/spack/spack/files/6393803/spack-build-02-configure-out.txt)
* [spack-build-01-autoreconf-out.txt](https://github.com/spack/spack/files/6393804/spack-build-01-autoreconf-out.txt)
* [config.log](https://github.com/spack/spack/files/6393806/config.log)

@lrknox as a maintainer of hdf package.
No maintainers found for libtirpc (which I suspect is the real culprit)

### General information

- [x ] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x ] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x ] I have uploaded the build log and environment files
- [x ] I have searched the issues of this repo and believe this is not a duplicate

**Comment by scheibelp** [comment]
If it is expected that all dependents of `libtirpc` will `#include <netconfig.h>` rather than `#include <libtirpc/netconfig.h>`, then I think the way to do this is to implement `def headers` in `libtirpc`:

```
    @property
    def headers(self):
        hdrs = find_all_headers(self.prefix.include, recursive=True)
        return hdrs or None
```

In that case, Spack will add `-I` options that include the full path to each `.h` file (whereas normally it would just add `-I<prefix>/include/`). 

Would you be able to try that?

**Comment by payerle** [comment]
I tried that just now (I think I tried something like that before submitting this ticket), and it failed in the same way (config.log test checking for xdr.h complains it cannot find rpc/types.h, etc.).  
(I removed the recursive=True flag, as find_all_headers does not allow for that, but does pass
recursive=True when it calls find_headers().)

There was no explicit -I for libtirpc in the configure gcc command, and spack-build-env.txt shows
\$prefix/include in SPACK_INCLUDE_DIRS, but not \$prefix/include/tirpc

Indeed, I used spack python with the following script:

```console
import spack
import llnl.util.filesystem as fs
libtirpc_prefix='/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4'

hdrs1 = fs.find_all_headers(libtirpc_prefix)
print("Using prefix")
print(hdrs1.cpp_flags)
print(hdrs1.directories)

hdrs2 = fs.find_all_headers("{}/include".format(libtirpc_prefix))
print("Using prefix/include")
print(hdrs2.cpp_flags)
print(hdrs2.directories)

hdrs3 = fs.find_all_headers("{}/include/tirpc".format(libtirpc_prefix))
print("Using prefix/include/tirpc")
print(hdrs3.cpp_flags)
print(hdrs3.directories)

print("hdrs1 == hdrs2\n")
print(hdrs1 == hdrs2)
print("hdrs1 == hdrs3\n")
print(hdrs1 == hdrs3)
print("hdrs2 == hdrs3\n")
print(hdrs2 == hdrs3)
```
and running that through ```spack python``` returned
```console
Using prefix
-I/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4/include
['/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4/include']
Using prefix/include
-I/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4/include
['/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4/include']
Using prefix/include/tirpc
-I/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4/include
['/software/spack-software/2021.04.28/linux-rhel8-skylake_avx512/gcc-8.4.0/libtirpc-1.2.6-on6c3p6qdpmzmffh3pmatghj2rkzotx4/include']
hdrs1 == hdrs2

True
hdrs1 == hdrs3

True
hdrs2 == hdrs3

True

```
I.e., I tried running  find_all_headers with just \$prefix, \$prefix/include, and \$prefix/include/tirpc, and in all cases it returned equivalent objects with cpp_path giving just \$prefix/include.

So the suggested redefinition of headers does not seem to be doing the trick, and I am at a loss as to how to get spack to alter the reported include directories for libtirpc.  Any help would be appreciated.

I could patch the install method to add symlinks, but I think modifying the header settings used by spack would be neater.

**Comment by scheibelp** [comment]
Thanks for that! Looking through the code I realized we take steps to "stop early" at `include/` since that is often what dependents want (but not the right thing to do here). When overriding headers, you can explicitly set the directories:

```
    @property
    def headers(self):
        hdrs = find_all_headers(self.prefix.include, recursive=True)
        if hdrs:
            hdrs.directories = [self.prefix.include.libtirpc]
            return hdrs
```

Up to this point it seems we haven't had use for functionality that automatically creates `-I` directives from all directories containing header files (hence why you have to explicitly name the directory in the example above). Spack could be modified though to provide this like:

```
    @property
    def headers(self):
        hdrs = find_all_headers(self.prefix.include, recursive=True)
        if hdrs:
            hdrs.directories = hdrs.all_directories # Normally a HeaderList truncates directories after 'include/', this does not
            return hdrs
```

**Comment by payerle** [comment]
OK, that worked.  I submitted a pull request to use both include and include/tirpc (just in case anything was expecting tirpc to be in the path).

**Comment by scheibelp** [comment]
Closing as fixed by #23332
